Implement osu target mod

hlysine opened the pull request:
Closes #8549

Depends on:
- [x] #13626
- [x] #13645

**To do:**
- [x] Add a seed for random circle generation (similar to #12568)
- [x] ~Switch to results screen after failing~ Show the fail overlay as usual instead
- [x] Replace sample for metronome
- [x] Match jump distance and circle scaling to those in cutting-edge
- [x] Avoid overlapping circles

**Differences with cutting-edge:**
- Cutting-edge uses the first hit object in the map as the first beat, which feels strange if the map starts on an offbeat. This implementation uses the start time of the first timing point (or the one that is active at the first hit object) instead. Example: the first few seconds of https://osu.ppy.sh/beatmapsets/1146827#osu/2394642
- Cutting-edge does not generate circles during breaks. This implementation does not generate circles from the start of a break to the start of the first hit object after the break. This makes the mod work well with compilation maps.
- Cutting-edge groups every 8 circles into a combo. This implementation respects the combos in the original map.

hlysine committed:
Implemented osu target mod

hlysine committed:
Refactored ApplyToBeatmap

smoogipoo commented:
I'm not sure that this should be implemented as a mod (as opposed to an entire ruleset), with it being so different to the point that even the sprites used should be different.

peppy commented:
I think at most the sprites would be replaced via a skin implementation (maybe restoring what stable did when we have more advanced skinning in place).

Is fine to go ahead without any visual replacements for now, for sure.

hlysine committed:
Added seed setting, fixed mod ignoring IncreaseFirstObjectVisibility

Henry-YSLin commented:
The seed setting is just tentative. I'm guessing it won't work for multi if the user doesn't set an explicit seed. My idea to fix this is to add a new IApplicable_ type that allows the mod to read `MultiplayerRoomSettings` from `MultiplayerClient` and use `PlaylistItemId` as the default seed. Does that sound alright?

PercyDan54 commented on osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs:
```diff
@@ -1,19 +1,496 @@
+            healthProcessor.FailConditions += (_, result)
+                => result.Type.AffectsCombo()
+                && !result.IsHit;
+        }
```
Why not `IApplicableFailOverride`

hlysine committed:
Display results after fail

peppy commented:
Have you checked how other mods handle seeing (see `ModRandom`, which recently got this added at an abstract level)? You can likely just derive `ModRandom` for your implementation, or if that isn't possible for some reason, abstract the seed logic out of `ModRandom` into a new `ModWithSeed` base class.

hlysine committed:
Merge branch 'ppy:master' into osu-target-mod

hlysine committed:
Extract seed setting contorl to IHasSeed

Henry-YSLin commented:
That was exactly what I was thinking, but `OsuModTarget` is already inheriting from `ModWithVisibilityAdjustment`, so the closest I can do is an `IHasSeed` interface.

hlysine committed:
Code cleanup

hlysine committed:
Break ApplyToBeatmap into subroutines

hlysine committed:
Improved guesstimations; fixed hit samples

peppy commented:
If it helps getting things accurate, here's the source for the original target practice mod. Note that it isn't necessary to match it 1:1 as it's only present on cutting-edge and not really considered an in-stone implementation.

https://gist.github.com/peppy/ac4b905861d674e17e3e1a88aa7cc05e

hlysine committed:
Increase distance between combos; pull circles closer to center

hlysine committed:
Dim circles instead of fade; improved hit samples; changed jump distance to be closer to cuttingedge

hlysine committed:
Make circles light up 1 beat length before start time

hlysine committed:
Animate circles undimming

peppy commented on osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs:
```diff
@@ -139,7 +139,7 @@ protected override void ApplyNormalVisibilityState(DrawableHitObject drawable, A
                 var avgColour = colour.AverageColour.Linear;
                 drawable.FadeColour(new Color4(avgColour.R * 0.45f, avgColour.G * 0.45f, avgColour.B * 0.45f, avgColour.A))
-                        .Then().Delay(h.TimePreempt - controlPointInfo.TimingPointAt(h.StartTime).BeatLength).FadeColour(colour);
+                        .Then().Delay(h.TimePreempt - controlPointInfo.TimingPointAt(h.StartTime).BeatLength - 96).FadeColour(colour, 96);
```
constants like this need to be extracted out to `private const` fields, with xmldoc where the name isn't already explaining what they mean.

hlysine committed:
Extract constants and add xmldoc

hlysine committed:
Take circle radius into account when clamping hit objects to playfield

hlysine committed:
Avoid overlapping with recent circles

hlysine committed:
Guard against edge cases

hlysine committed:
Remove the circle size buff

hlysine committed:
Marked target mod and traceable mod as incompatible; extracted playfield clamping logic

Nothing is visible when target mod and traceable mod are enabled together.

Henry-YSLin commented:
Is there a sample for the metronome in this mod? I can't find one in osu-resources.

peppy commented on osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs:
```diff
@@ -51,14 +51,26 @@ public class OsuModTarget : ModWithVisibilityAdjustment, IApplicableToDrawableRu
+        private const float max_base_distance = 250f;
+
+        /// <summary>
+        ///     The maximum allowed jump distance after multipliers are applied
```
```suggestion
        /// The maximum allowed jump distance after multipliers are applied
```

Incorrect spacing, also applies to all other xmldoc.

Henry-YSLin replied:
Already fixed. I had the wrong coding style setting in Rider.

hlysine committed:
Switched to a more reasonable sample sound for now

Henry-YSLin commented:
The only unimplemented mechanic now is the shattering of target sprites when they are hit, which should come together with the visual replacements later. Aside from that, this mod should be ready for a full review.

bdach commented:
Just by scrolling this diff I can see that this pull is doing several things at once, namely:

* implementing the target mod itself
* adding a new mod API in `IApplicableFailOverride.DisplayResultsOnFail` (and seemingly setting it to `false` for anything that isn't the target mod)
* adding a new `IHasSeed` interface and a control for it

And partly due to the above, has grown to XL size without necessarily needing to and has several unresolved merge conflicts. Unsure how much of a fuss to make about this but splitting this into chunks would definitely make this more digestible given that the actual mod implementation itself is almost 500 lines on its own.

Henry-YSLin commented:
Noted. I will open new PRs for `IApplicableFailOverride.DisplayResultsOnFail` and `IHasSeed`. After they are merged, I will come back to target mod and hopefully make the diff less intimidating.

bdach commented:
As far as I can see the rejected changes from #13616 were not originally a part of this pull. Were they required in some capacity? They don't seem to be.

Henry-YSLin commented:
No, they weren't required. I just thought the naming was confusing while I was implementing support for `IApplicableFailOverride.DisplayResultsOnFail` in the newly rewritten `Player`.

hlysine committed:
Revert "Display results after fail"

This commit reverts 7815b3c7

hlysine committed:
Merge branch 'master' into osu-target-mod

# Conflicts:
#	osu.Game.Rulesets.Osu/Mods/OsuModTraceable.cs
#	osu.Game/Rulesets/Mods/IHasSeed.cs

hlysine committed:
Implement IMutateApproachCircles

hlysine committed:
Reordered things and added regions

bdach commented on osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs:
```diff
@@ -1,19 +1,528 @@
+        {
+            var startTime = origHitObjects.First().StartTime;
+            var endObj = origHitObjects.Last();
+            var endTime = endObj switch
+            {
+                Slider slider => slider.EndTime,
+                Spinner spinner => spinner.EndTime,
+                _ => endObj.StartTime
+            };
```
Should be `endObj.GetEndTime()` (it's an extension method). Also we don't use switch expressions in this repository.

bdach commented on osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs:
```diff
@@ -1,19 +1,528 @@
+                               .Where(x => Precision.AlmostBigger(endTime, x.Time))
+                               .SelectMany(tp =>
+                               {
+                                   var tpBeats = new List<double>();
+                                   var currentTime = tp.Time;
+
+                                   while (Precision.AlmostBigger(endTime, currentTime) && beatmap.ControlPointInfo.TimingPointAt(currentTime) == tp)
+                                   {
+                                       tpBeats.Add(Math.Floor(currentTime));
+                                       currentTime += tp.BeatLength;
+                                   }
+
+                                   return tpBeats;
```
* This should really be a method rather than an inline linq lambda. And one with preferably longer variable names.
* This method of generating beats is susceptible to drifting out of sync due to accumulating floating point error. To avoid that this should be done via multiplication, e.g. `startTime + i * beatLength` where `i` increments up until the entire expression exceeds `endTime`.

bdach commented on osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs:
```diff
@@ -1,19 +1,528 @@
+                               .Where(x => Precision.AlmostBigger(x, startTime))
+                               // Remove beats during breaks
+                               .Where(x => !beatmap.Breaks.Any(b =>
+                                   Precision.AlmostBigger(x, b.StartTime)
+                                   && Precision.AlmostBigger(origHitObjects.First(y => Precision.AlmostBigger(y.StartTime, b.EndTime)).StartTime, x)
```
This needs to be split out more, I can't follow what this is attempting to do at all. Especially not with variable names of the `b`, `y`, `x` kind.

Also all of these `Precision` calls around are using the default precision of `1e-3` for floats which seems too little for timing objects. It should probably be `1` (millisecond).

bdach commented on osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs:
```diff
@@ -1,19 +1,528 @@
+                    || !Precision.AlmostBigger(s.EndTime, time))
+                    return false;
+
+                return Precision.AlmostEquals((time - s.StartTime) % s.SpanDuration, 0);
```
The modulo op on floats is dangerous, especially if we're talking about precision. With fractions that are fundamentally unrepresentable using floating point arithmetic you can get unexpected results. Example:

```
12.1 % 0.1 = 0.09999999999999898
 0.2 % 0.1 = 0
-0.1 % 0.1 = -0
-0.3 % 0.1 = -0.09999999999999998
```

So doing things this way may not prove to be a good idea. You may want to consider doing division here and checking the fractional part of the dividend or something of the sort. Compare mania's `DistanceObjectPatternGenerator.nodeSamplesAt()`.

bdach commented on osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs:
```diff
@@ -1,19 +1,528 @@
+            {
+                if (Precision.AlmostEquals(time, x.StartTime)) return true;
+
+                if (!(x is Slider s))
```
Should really be using `IHasPathWithRepeats` for this sort of thing.

bdach commented on osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs:
```diff
@@ -1,19 +1,528 @@
+                {
+                    while (lastSampleIdx < origHitObjects.Count && origHitObjects[lastSampleIdx].StartTime <= x.StartTime)
+                        lastSampleIdx++;
+                    lastSampleIdx--;
+
+                    if (lastSampleIdx < 0 && lastSampleIdx >= origHitObjects.Count) continue;
```
What's the decrement for? And the `lastSampleIdx >= Count` check, too?

Henry-YSLin replied:
The decrement was originally needed because the `while` always loops past the target hit object. I've reworked the logic a bit so that it is no longer needed.

bdach commented on osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs:
```diff
@@ -1,19 +1,528 @@
+            // First follow the combo indices in the original beatmap
+            hitObjects.ForEach(x =>
+            {
+                var origObj = origHitObjects.FindLast(y => Precision.AlmostBigger(x.StartTime, y.StartTime));
+                x.ComboIndex = origObj?.ComboIndex ?? 0;
```
These linq expressions direly need actual explanations for what they're doing. I think this is trying to fetch the closest object preceding `x`? I can't get to that from "follow the combo indices in the original beatmap".

bdach commented on osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs:
```diff
@@ -1,19 +1,528 @@
+                var origObj = origHitObjects.FindLast(y => Precision.AlmostBigger(x.StartTime, y.StartTime));
+                x.ComboIndex = origObj?.ComboIndex ?? 0;
+            });
+            // Then reprocess them to ensure continuity in the combo indices and add indices in current combo
+            var combos = hitObjects.GroupBy(x => x.ComboIndex).ToList();
```
If this is going to be done, wouldn't it be simpler to get a list of all start times of objects that have `NewCombo = true`, and then just iteratively reindex hitobjects up to the next new combo?

Henry-YSLin replied:
I have tried this approach, but the code is much more complex because there are cases where an entire combo in the original map gets skipped over. See the newly added comments for details.

bdach commented on osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs:
```diff
@@ -1,19 +1,528 @@
+        {
+            if (hitObjects.Count == 0) return;
+
+            var rng = new Random(Seed.Value.GetValueOrDefault());
```
I'm pretty sure this isn't doing what is intended. If `Seed.Value` is null, then calling `.GetValueOrDefault()` returns 0. So this is not in fact using some random seed, it's using exactly seed 0 every time.

`new Random(0)` is not equivalent to `new Random()`. The latter [uses tick count for initialisation](https://referencesource.microsoft.com/#mscorlib/system/random.cs,52).

Henry-YSLin replied:
`Seed.Value` can never be null at this point because it has already been replaced by a random value if the user doesn't specify a seed. I should move this to a better place to avoid the confusion.

bdach replied:
There should be an assertion that the seed is not null in that case, instead of falling back to a default that is never supposed to actually happen

bdach commented on osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs:
```diff
@@ -1,19 +1,528 @@
+
+            float nextSingle(float max = 1f) => (float)(rng.NextDouble() * max);
+
+            var direction = MathHelper.TwoPi * nextSingle();
```
Please don't reference osuTK constants for 2*pi. If it is really intended to have it with single precision, you can use a (method-local) const.

bdach commented on osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs:
```diff
@@ -1,19 +1,528 @@
+                    if (tryCount > 0) direction = MathHelper.TwoPi * nextSingle();
+
+                    var relativePos = new Vector2(
+                        distance * (float)Math.Cos(direction),
+                        distance * (float)Math.Sin(direction)
+                    );
```
Use `MathF` methods rather than `(float)Math`

bdach commented on osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs:
```diff
@@ -1,19 +1,528 @@
+        private bool isOverlappingWithRecent(IReadOnlyList<OsuHitObject> hitObjects, int idx)
+        {
+            var target = hitObjects[idx];
+            return hitObjects.SkipLast(hitObjects.Count - idx).TakeLast(overlap_check_count)
+                             .Any(h => Vector2.Distance(h.Position, target.Position) < target.Radius * 2);
```
Why is this operating on the entire list and iindirectly iterating over it via linq rather than taking a `Span` of the few objects that are actually needed here or something of the sort?

Henry-YSLin replied:
I've refactored the method, but I'm not sure if you can take a `Span` from a `List`, so the method is accepting an `IEnumerable` for now.

bdach commented on osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs:
```diff
@@ -1,19 +1,528 @@
+        /// Determines the position of the current hit object relative to the previous one.
+        /// </summary>
+        /// <returns>The position of the current hit object relative to the previous one</returns>
+        private Vector2 getRotatedVector(Vector2 prevPosChanged, Vector2 posRelativeToPrev)
```
These are lifted verbatim from `OsuModRandom` and probably should be split to a common helper.

Henry-YSLin replied:
This was originally on the to-do list but got forgotten somehow. I will open a new PR for this.

bdach commented:
The above review was mostly surface-level, without digging too much into the code. Readability improvements are direly needed; I think that linq has been (over)used in this code to the point of obfuscating logic.

hlysine committed:
Use `GetEndTime()` instead of a switch expression

hlysine committed:
Converted an inline lambda into a method (`getBeatsForTimingPoint`)

hlysine committed:
Converted an inline lambda into a method (`isInsideBreakPeriod`); moved `origHitObjects` to be a private class field

hlysine committed:
Avoid modulo when finding slider node index

hlysine committed:
Switch to `IHasPathWithRepeats`

hlysine committed:
Removed odd-looking decrement and checks in `addHitSamples`

hlysine committed:
Moved a misplaced `!`

hlysine committed:
Better explanation for `fixComboInfo`

hlysine committed:
Moved RNG initialisation to a better place

hlysine committed:
Avoid using osuTK constants; Use `MathF`

hlysine committed:
Refactor and rename `isOverlappingWithRecent`

hlysine committed:
Merge branch 'master' into osu-target-mod

hlysine committed:
Added comments

hlysine committed:
Directly fade to gray instead of computing the color values

hlysine committed:
Minor fixes in comments

hlysine committed:
Increase acceptable difference for Precision calls

hlysine committed:
Merge branch 'master' into osu-target-mod

hlysine committed:
Use `OsuHitObjectGenerationUtils`

hlysine committed:
Use `IHasRepeats` instead of `IHasPathWithRepeats`

hlysine committed:
Split part of `addHitSamples` to a subroutine

bdach committed:
Merge branch 'master' into osu-target-mod

bdach committed:
Change seed control type in line with changes

bdach commented on osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs:
```diff
@@ -1,19 +1,521 @@
+
+                    tryCount++;
+                    if (tryCount % 10 == 0) distance *= 0.9f;
+                } while (distance >= obj.Radius * 2 && checkForOverlap(hitObjects.SkipLast(hitObjects.Count - i).TakeLast(overlap_check_count), obj));
```
Bit uneasy about the lack of a `tryCount` check in this termination condition, but I do suppose this has to converge in a finite number of iterations because of the constraints placed on `distance` (must be less than `distance_cap`, is geometrically shrunk every iteration).

Also I imagine this `checkForOverlap`'s overhead will be nasty (quadratic) due to how many objects are being skipped using linq. Something like a limited size queue would work better here.

Henry-YSLin replied:
The `SkipLast().TakeLast()` combo should run in constant time as long as the length of the resulting enumeration is constant (`overlap_check_count` in this case), so it shouldn't be a huge worry.

bdach commented on osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs:
```diff
@@ -1,19 +1,521 @@
+
+        private OsuHitObject getClosestHitObject(List<OsuHitObject> hitObjects, double time)
+        {
+            var closestIdx = hitObjects.FindLastIndex(h => h.StartTime < time);
```
`precedingIndex`

Henry-YSLin replied:
A question: should I use `almostBigger` for `h.StartTime < time` here? I know that `h.StartTime` cannot be close to `time` but the reason for that lies in `getSamplesAtTime` and can take quite a while to comprehend.

bdach replied:
I'm not sure how `<` vs `almostBigger` matters here. So plain `<` for readability, probably.

bdach commented on osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs:
```diff
@@ -1,19 +1,521 @@
+            double spanDuration = curve.Duration / curve.SpanCount();
+            double nodeIndex = timeSinceStart / spanDuration;
+
+            if (almostEquals(nodeIndex - Math.Round(nodeIndex), 0))
```
```suggestion
            if (almostEquals(nodeIndex, Math.Round(nodeIndex)))
```

bdach reviewed (dismissed):
better readability-wise, but still needs work

Henry-YSLin commented on osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs:
```diff
@@ -1,19 +1,521 @@
+            // Copy combo indices from the closest preceding object
+            hitObjects.ForEach(obj =>
+            {
+                var closestOrigObj = origHitObjects.FindLast(y => almostBigger(obj.StartTime, y.StartTime));
```
Yes, this is intentional. Hit objects should copy combo indices from an original HO at the same start time if possible, so `obj.StartTime == closestOrigObj.StartTime` is expected.

Henry-YSLin replied:
I have changed some calls to use `definitelyBigger`, mostly for readability's sake. I already considered the equality cases when using `almostBigger`, which was why some calls were strangely inverted.

Henry-YSLin commented on osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs:
```diff
@@ -1,19 +1,521 @@
+
+                var distance = maxComboIndex == 0
+                    ? (float)obj.Radius
+                    : map(obj.ComboIndex, 0, maxComboIndex, (float)obj.Radius, max_base_distance);
```
Not sure about this. Using `Interpolation.Lerp` would result in very similar code to the implementation of `map`, and `map` is a pretty standard name for a function that maps a value from one range to another imo.

bdach replied:
`map` is an overly generic name. You can map almost anything to anything. java/functional languages use that as a synonym to C# `.Select()`. `mapRange` would be better by me if you insist on using that.

hlysine committed:
Several requested changes

- Rename `origHitObjects` to `originalHitObjects`
- Use `Value` instead of `GetValueOrDefault()`
- Remove `endObj`
- Added comments
- Rename `closestIdx` to `precedingIndex`
- Changed an `almostEquals` call

hlysine committed:
Added `definitelyBigger`

hlysine committed:
Prevent multiple enumeration in `checkForOverlap`

hlysine committed:
Merge branch 'master' into osu-target-mod

hlysine committed:
Use `IHidesApproachCircles`

bdach commented:
I could be fine with this diff, but there is one more matter to resolve - `TestSceneModSelectOverlay.TestOsuMods` is failing because it was relying on the target mod being unimplemented. Unsure whether to just nuke the offending assert, or try to add coverage for unimplemented mods being unselectable in another manner.

hlysine committed:
Rename `map` to `mapRange`

Henry-YSLin commented:
Creating a mock ruleset with a mock unimplemented mod is the safest way to test this imo. Not sure where to put these mock classes though.

bdach commented:
Having those as inner classes of the test scene may be fine. It'd probably best be done in another separate pull, because this one is big enough on its own.

hlysine committed:
Merge branch 'master' into osu-target-mod

Henry-YSLin commented:
@bdach A friendly ping: the test failure has been fixed. The current CI failure should be unrelated to this PR.

bdach commented:
I'll dismiss my review, but this has been through so many rounds of review that I'm not super willing to merge without a double check from someone else.

peppy commented:
I think overall the code quality seems fine for this, but I'm curious where you got the dim/fade behaviour from. This dim is something that needs to be implemented at a ruleset level, so I'm not sure I agree with it being done locally, but even if it is done here, the dim values seem completely random (grey 0.45).

Henry-YSLin commented:
Not sure how this should be implemented at a ruleset level. Isn't the dim behaviour specific to this mod? 

As for the dim values, I got that from the cutting-edge implementation:
```csharp
Color col1 = new Color((byte)Math.Max(0, Colour.R * 0.45F), (byte)Math.Max(0, Colour.G * 0.45F), (byte)Math.Max(0, Colour.B * 0.45F), 255);
```

peppy commented:
Right, thanks. I somehow missed that code. I guess this is actually applying some heavier dim than usual. The discrepancy is likely due to the way colour application is different in lazer and stable.

peppy commented:
The fade in timings don't match stable's. It's important they do, as they are aligned with the beat (fade is exactly one beat earlier, so you have some visual cues). ie. `Preempt` should not be used at all.

peppy commented:
The issue is you're using `PreemptTime` in the absolute sequence. You should be starting your sequence from `hitObject.StartTime - beatLength - 96` or something like that.

Henry-YSLin commented:
But the entire sequence should be
- apply fade at `hitObject.StartTime - hitObject.TimePreempt`
- gradually remove fade from `hitObject.StartTime - beatLength - 96` to `hitObject.StartTime - beatLength`

Shouldn't it start at `hitObject.StartTime - hitObject.TimePreempt` then?

peppy commented:
Sure, that sounds about right.

Henry-YSLin commented:
> Shouldn't it start at `hitObject.StartTime - hitObject.TimePreempt` then?

That is the current behaviour. I was just wondering whether 96ms is accurate since it feels off in gameplay.

peppy commented:
Okay, it's just written so weirdly it looks wrong. I'll refactor it.

peppy committed:
Adjust transform logic to hopefully be a bit easier to parse

peppy commented:
One remaining issue I've noticed is the metronome plays twice near the first object, and doesn't play a countdown before the first object.

Henry-YSLin commented:
The metronome playing twice at the beginning is caused by `BeatSyncedContainer` always firing `OnNewBeat` when the current time of the beatmap track is 0 and the beat index is negative. Is this intended behaviour? If not, it should probably be fixed in a separate PR.

As for the countdown, cutting-edge seems to extend the intro time to fit one bar of countdown for the metronome. No idea how to do that just yet.

peppy committed:
Use a slightly more appropriate metronome sound

peppy committed:
Fix metronome playing during intro time

peppy committed:
Merge branch 'master' into osu-target-mod

peppy commented:
I've fixed the remaining issues in this PR. Should be good to go.

My changes were minimal so I'm just going to get this one in. The other PR #13890 can come as a follow-up (I will apply the changes required to this mod in that PR.

bdach commented on osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs:
```diff
@@ -1,19 +1,554 @@
+            using (drawable.BeginAbsoluteSequence(startTime - controlPointInfo.TimingPointAt(startTime).BeatLength - undim_duration))
+                drawable.FadeColour(Color4.White, undim_duration);
+
+            // Remove approach circles
+            circle.ApproachCircle.Hide();
```
Should this be out here like this, rather than inside the `startTime - preempt` absolute sequence? All other usages have such hide logic inside a transform sequence.

peppy replied:
I think due to what this is doing it's fine as-is, but let's adjust it for consistency.

bdach reviewed (commented):
Skimmed through @peppy's latest changes and all except maybe one look OK.

peppy committed:
Move approach circle hiding within `BeginAbsoluteSequence`

hlysine committed:
Code formatting fixes

bdach committed:
Merge branch 'master' into osu-target-mod
